Added the unit tests.

$ python EventAwareLumiBased_t.py.......
----------------------------------------------------------------------
Ran 7 tests in 0.110s

OK


If I revert back  commit 1, the new testG_LumiMask gives:

$ python EventAwareLumiBased_t.py......F
======================================================================
FAIL: testG_LumiMask (__main__.EventAwareLumiBasedTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "EventAwareLumiBased_t.py", line 591, in testG_LumiMask
    self.assertEqual(jobs[0]['mask'].getRunAndLumis(), {1L: [[10L, 14L]], 2L: [[20L, 21L]], 4L: [[40L, 40L]]})
AssertionError: {1: [[10, 14]], 2: [[20, 21]]} != {1L: [[10L, 14L]], 2L: [[20L, 21L]], 4L: [[40L, 40L]]}

----------------------------------------------------------------------
Ran 7 tests in 0.131s

FAILED (failures=1)

This is same to what I described about my CRAB3 test task. In testG_LumiMask, each job should analyze 8 lumis (each lumi has 100 events and I set events_per_job to 850). But job 1 got 7 lumis (lumi 4L: [40L, 40L] is given to job 2 instead of job 1). This is because when switching from fileB to fileC, lumisInJob = 7 while f['lumiCount'] = 4 (f = fileB). So min(lumisInJob, f['lumiCount']) = f['lumiCount'] = 4 which means that the estimate of the average number of events in job 1 is 100 * 5 (from fileA) + 100 * 4 (from fileB) = 900 instead of 100 * 5 + 100 * 2 (only 2 from the 4 in fileB are good lumis) = 700. With the wrong estimate of avgEvts = 900 there is no room for another lumi in job 1; with the correct estimate of avgEvts = 700 there is room for 1 more lumi.

--- a/src/python/WMCore/JobSplitting/EventAwareLumiBased.py
+++ b/src/python/WMCore/JobSplitting/EventAwareLumiBased.py
@@ -41,6 +41,7 @@ class EventAwareLumiBased(JobFactory):
 
         avgEventsPerJob = int(kwargs.get('events_per_job', 5000))
         eventLimit      = int(kwargs.get('max_events_per_lumi', 20000))
+        totalEvents     = int(kwargs.get('total_events', 0))
         splitOnFile     = bool(kwargs.get('halt_job_on_file_boundaries', True))
         ignoreACDC      = bool(kwargs.get('ignore_acdc_except', False))
         collectionName  = kwargs.get('collectionName', None)
@@ -138,7 +139,9 @@ class EventAwareLumiBased(JobFactory):
         firstLumi      = None
         lastRun        = None
         lumisInJob     = 0
+        totalAvgEventCount = 0
         currentJobAvgEventCount = 0
+        stopTask = False
         for location in locationDict:
 
             # For each location, we need a new jobGroup
@@ -275,9 +278,15 @@ class EventAwareLumiBased(JobFactory):
                         lastLumi = lumi
                         stopJob = False
                         lastRun = run.run
+                        totalAvgEventCount += f['avgEvtsPerLumi']
 
                         if self.currentJob and not f in self.currentJob['input_files']:
                             self.currentJob.addFile(f)
+
+                        # We stop here if there are more total events than requested.
+                        if totalEvents > 0 and totalAvgEventCount >= totalEvents:
+                            stopTask = True
+                            break
 
                     if firstLumi != None and lastLumi != None:
                         # Add this run to the mask
@@ -290,7 +299,16 @@ class EventAwareLumiBased(JobFactory):
                         firstLumi = None
                         lastLumi = None
 
+                    if stopTask:
+                        break
+
                 if not splitOnFile:
                     currentJobAvgEventCount += f['avgEvtsPerLumi'] * lumisInJobInFile
+
+                if stopTask:
+                    break
+
+            if stopTask:
+                break
 
         return

--- a/test/python/WMCore_t/JobSplitting_t/EventAwareLumiBased_t.py
+++ b/test/python/WMCore_t/JobSplitting_t/EventAwareLumiBased_t.py
@@ -215,6 +215,76 @@ class EventAwareLumiBasedTest(unittest.TestCase):
         # in it.
         self.assertEqual(jobs[0]['mask'].getRunAndLumis(), {0L: [[0L, 2L]]})
         self.assertEqual(jobs[1]['mask'].getRunAndLumis(), {0L: [[3L, 4L]]})
+
+        # Test total_events limit. (The algorithm cuts off after the lumi that
+        # brings the total average event count over -or equal to- total_events.)
+        testSubscription = self.createSubscription(nFiles = 5, lumisPerFile = 5, twoSites = False)
+        jobFactory = splitter(package = "WMCore.DataStructs",
+                              subscription = testSubscription)
+
+        jobGroups = jobFactory(halt_job_on_file_boundaries = False,
+                               splitOnRun = False,
+                               events_per_job = 60,
+                               total_events = 10,
+                               performance = self.performanceParams)
+
+        self.assertEqual(len(jobGroups), 1)
+        jobs = jobGroups[0].jobs
+        self.assertEqual(len(jobs), 1)
+        self.assertEqual(jobs[0]['mask'].getRunAndLumis(), {0L: [[0L, 0L]]})
+
+        testSubscription = self.createSubscription(nFiles = 5, lumisPerFile = 5, twoSites = False)
+        jobFactory = splitter(package = "WMCore.DataStructs",
+                              subscription = testSubscription)
+
+        jobGroups = jobFactory(halt_job_on_file_boundaries = False,
+                               splitOnRun = False,
+                               events_per_job = 60,
+                               total_events = 179,
+                               performance = self.performanceParams)
+
+        self.assertEqual(len(jobGroups), 1)
+        jobs = jobGroups[0].jobs
+        self.assertEqual(len(jobs), 3)
+        self.assertEqual(jobs[0]['mask'].getRunAndLumis(), {0L: [[0L, 2L]]})
+        self.assertEqual(jobs[1]['mask'].getRunAndLumis(), {0L: [[3L, 4L]], 1L: [[5L, 5L]]})
+        self.assertEqual(jobs[2]['mask'].getRunAndLumis(), {1L: [[6L, 8L]]})
+
+        testSubscription = self.createSubscription(nFiles = 5, lumisPerFile = 5, twoSites = False)
+        jobFactory = splitter(package = "WMCore.DataStructs",
+                              subscription = testSubscription)
+
+        jobGroups = jobFactory(halt_job_on_file_boundaries = False,
+                               splitOnRun = False,
+                               events_per_job = 60,
+                               total_events = 180,
+                               performance = self.performanceParams)
+
+        self.assertEqual(len(jobGroups), 1)
+        jobs = jobGroups[0].jobs
+        self.assertEqual(len(jobs), 3)
+        self.assertEqual(jobs[0]['mask'].getRunAndLumis(), {0L: [[0L, 2L]]})
+        self.assertEqual(jobs[1]['mask'].getRunAndLumis(), {0L: [[3L, 4L]], 1L: [[5L, 5L]]})
+        self.assertEqual(jobs[2]['mask'].getRunAndLumis(), {1L: [[6L, 8L]]})
+
+        testSubscription = self.createSubscription(nFiles = 5, lumisPerFile = 5, twoSites = False)
+        jobFactory = splitter(package = "WMCore.DataStructs",
+                              subscription = testSubscription)
+
+        jobGroups = jobFactory(halt_job_on_file_boundaries = False,
+                               splitOnRun = False,
+                               events_per_job = 60,
+                               total_events = 181,
+                               performance = self.performanceParams)
+
+        self.assertEqual(len(jobGroups), 1)
+        jobs = jobGroups[0].jobs
+        self.assertEqual(len(jobs), 4)
+        self.assertEqual(jobs[0]['mask'].getRunAndLumis(), {0L: [[0L, 2L]]})
+        self.assertEqual(jobs[1]['mask'].getRunAndLumis(), {0L: [[3L, 4L]], 1L: [[5L, 5L]]})
+        self.assertEqual(jobs[2]['mask'].getRunAndLumis(), {1L: [[6L, 8L]]})
+        self.assertEqual(jobs[3]['mask'].getRunAndLumis(), {1L: [[9L, 9L]]})
+
         return
 
     def testC_FileSplitNoHardLimit(self):
@@ -279,6 +349,27 @@ class EventAwareLumiBasedTest(unittest.TestCase):
         jobs = jobGroups[0].jobs
         self.assertEqual(len(jobs), 25, "There should be 5 jobs")
 
+        # Test total_events limit. (The algorithm cuts off after the lumi that
+        # brings the total average event count over -or equal to- total_events.)
+        testSubscription = self.createSubscription(nFiles = 5, lumisPerFile = 3, twoSites = False,
+                                                   nEventsPerFile = 300)
+        jobFactory = splitter(package = "WMCore.DataStructs",
+                              subscription = testSubscription)
+        jobGroups = jobFactory(halt_job_on_file_boundaries = True,
+                               splitOnRun = True,
+                               events_per_job = 250,
+                               total_events = 750,
+                               performance = self.performanceParams)
+        self.assertEqual(len(jobGroups), 1, "There should be only one job group")
+        jobs = jobGroups[0].jobs
+        self.assertEqual(len(jobs), 5, "There should be 5 jobs")
+        self.assertEqual(jobs[0]['mask'].getRunAndLumis(), {0L: [[0L, 1L]]})
+        self.assertEqual(jobs[1]['mask'].getRunAndLumis(), {0L: [[2L, 2L]]})
+        self.assertEqual(jobs[2]['mask'].getRunAndLumis(), {1L: [[3L, 4L]]})
+        self.assertEqual(jobs[3]['mask'].getRunAndLumis(), {1L: [[5L, 5L]]})
+        self.assertEqual(jobs[4]['mask'].getRunAndLumis(), {2L: [[6L, 7L]]})
+
+        return
 
     def testD_NoFileSplitNoHardLimit(self):
         """
@@ -473,6 +564,67 @@ class EventAwareLumiBasedTest(unittest.TestCase):
                           "The reason for the failure is not accurate")
 
         return
+
+    def testG_LumiMask(self):
+        """
+        _testG_LumiMask_
+
+        Test that we can use a lumi-mask to filter good runs/lumis.
+        """
+        splitter = SplitterFactory()
+
+        # Create 3 files with 100 events per lumi:
+        # - file1 with 1 run  of 8 lumis
+        # - file2 with 2 runs of 2 lumis each
+        # - file3 with 1 run  of 5 lumis
+        fileA = File(lfn = "/this/is/file1", size = 1000, events = 800)
+        fileB = File(lfn = "/this/is/file2", size = 1000, events = 400)
+        fileC = File(lfn = "/this/is/file3", size = 1000, events = 500)
+
+        lumiListA = []
+        for lumi in range(8):
+            lumiListA.append(10 + lumi)
+        fileA.addRun(Run(1, *lumiListA))
+        fileA.setLocation("somese.cern.ch")
+        lumiListB1 = []
+        lumiListB2 = []
+        for lumi in range(2):
+            lumiListB1.append(20 + lumi)
+            lumiListB2.append(30 + lumi)
+        fileB.addRun(Run(2, *lumiListB1))
+        fileB.addRun(Run(3, *lumiListB2))
+        fileB.setLocation("somese.cern.ch")
+        lumiListC = []
+        for lumi in range(5):
+            lumiListC.append(40 + lumi)
+        fileC.addRun(Run(4, *lumiListC))
+        fileC.setLocation("somese.cern.ch")
+
+        testFileset = Fileset(name = 'Fileset')
+        testFileset.addFile(fileA)
+        testFileset.addFile(fileB)
+        testFileset.addFile(fileC)
+
+        testSubscription = Subscription(fileset = testFileset,
+                                        workflow = self.testWorkflow,
+                                        split_algo = "EventAwareLumiBased",
+                                        type = "Processing")
+        jobFactory = splitter(package = "WMCore.DataStructs",
+                              subscription = testSubscription)
+
+        # Use a lumi-mask = {1: [[10,14]], 2: [[20,21]], 4: [[40,41]]}
+        jobGroups = jobFactory(halt_job_on_file_boundaries = False,
+                               splitOnRun = False,
+                               events_per_job = 850,
+                               runs = ['1', '2', '4'],
+                               lumis = ['10,14', '20,21', '40,41'],
+                               performance = self.performanceParams)
+
+        self.assertEqual(len(jobGroups), 1, "There should be only one job group")
+        jobs = jobGroups[0].jobs
+        self.assertEqual(len(jobs), 2, "Two jobs must be in the jobgroup")
+        self.assertEqual(jobs[0]['mask'].getRunAndLumis(), {1L: [[10L, 14L]], 2L: [[20L, 21L]], 4L: [[40L, 40L]]})
+        self.assertEqual(jobs[1]['mask'].getRunAndLumis(), {4L: [[41L, 41L]]})
 
 
 if __name__ == '__main__':

